torch/_functorch/aot_autograd.py

```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
         return t.view(t.shape)
     return t
 
-# This function takes in a forward fn, runs it, and (optionally) runs autograd to compute the joint.
-# When maybe_tangents is None, we only run the forward. Otherwise we run the "joint" forward + backward.
-# Preconditions:
-# - fn corresponds to the flattened user fw function, with duplicate inputs removed
-# - functionalization is turned on (and inputs are wrapped in functional tensors)
-# - Synthetic bases have been *removed* (we've taken views on them corresponding to the user argument views).
-# - primals_after_cloning are what we run our forward function on. It is identical to primals_before_cloning,
-#   except that every input we know will be mutated in the forward has been cloned.
-#   We run our forward on primals_after_cloning (potentially mutating some inputs), and then compute our gradients
-#   w.r.t. primals_before_cloning (so we properly capture the mutation in our gradient computation).
+def fn_no_extra_mutations(
+    fn: Callable,
+    meta: ViewAndMutationMeta,
+    keep_input_mutations: bool,
```

Comment:
It's supremely unclear to me why we need all these knobs.

1. I still think we should have handled metadata mutation in its own pass early on
2. I'm not sure why we need `keep_input_mutations` to toggle what outputs we return. In particular, it seems harmless to unconditionally NOT return the mutated inputs and figure the copy_ calls later (because remember, you can always access the original inputs to get the mutated versions!)

Reply:
From our last offline discussion:

(1) Agreed on handling metadata mutation earlier being nice (although there will be some mumbling around handling resize_(), and I probably won't deal with it in this PR)

(2) on the keep_input_mutation handling: it sounds like you're suggesting that all of the logic around "returning updated inputs as outputs" could be handled entirely by the functionalization layer. I'm not totally convinced we can do this: at some point when we call `autograd.grad()`, we need to potentially pass in the mutated inputs as `needed_outs` (and corresponding tangents for them), so that we properly autograd through the input mutation. This can't happen at the functionalization layer, because the fn that `create_functionalized_graph()` takes in is the entire joint function itself - it needs to be inside of that joint function that we pass mutated inputs into `autograd.grad`

Right now locally, I have a `prepare_for_autograd()` function that basically takes in a function, and does all of the intermediate work of (1) clone mutated inputs, (2) return mutated inputs, (3) return intermediate bases. I then pass that new function to a very simple `create_joint()` function, that runs the given fn and calls `autograd.grad()` on the output (along with a mask to know which outputs to grad).